pkg/monitor/sampler.go

```diff
@@ -38,27 +33,31 @@ func (s *sampler) run(ctx context.Context) {
 	defer ticker.Stop()
 	var lastInterval int = -1
 	for {
-		success := s.isAvailable()
+		previousSampleWasAvailable := s.isAvailable()
 		// the sampleFn may take a significant period of time to run.  In such a case, we want our start interval
 		// for when a failure started to be the time when the request was first made, not the time when the call
 		// returned.  Imagine a timeout set on a DNS lookup of 30s: when the GET finally fails and returns, the outage
 		// was actually 30s before.
 		startTime := time.Now().UTC()
-		condition, ok := s.sampleFn(success)
+		condition, currentSampleIsAvailable := s.sampleFn(previousSampleWasAvailable)
 		if condition != nil {
 			s.recorder.RecordAt(startTime, *condition)
 		}
 		if s.onFailing != nil {
 			switch {
-			case !success && ok:
+			case !previousSampleWasAvailable && currentSampleIsAvailable:
 				if lastInterval != -1 {
 					s.recorder.EndInterval(lastInterval, time.Now().UTC())
 				}
-			case success && !ok:
-				lastInterval = s.recorder.StartInterval(startTime, *s.onFailing)
+			case previousSampleWasAvailable && !currentSampleIsAvailable:
+				if condition != nil { // if an edge condition is provided, use this to load the interval
```

Comment:
This was missing before and it is why we consistently got "nil" as the error instead of something useful.